### ISSUE TITLE
Confusion Matrix for Test Set

### ISSUE THREAD

**Issue by fbagci** [ultralytics/yolov5#3535]
This question may seem simple, but since I couldn't find an answer or manual, I wanted to write it here.

I have trained Yolov5s with a public dataset in colab as in roboflow guide (https://www.youtube.com/watch?v=MdF6x6ZmLAY). I got the confusion matrix for the training set and the validation set automatically after training, but I could not get confusion matrix for test set. 
I tried manually to get the confusion matrix for my test set using test.py. I got an error like in the picture. The most important thing that measures model accuracy is the performance on test set isn't it? What issue am I missing here? Thanks.

![resim](https://user-images.githubusercontent.com/33219115/121241449-1fe19000-c8a4-11eb-9a77-12a239e43d25.png)

**Comment by glenn-jocher** [comment]
@fbagci test.py can operate on any split: train, val, test using the `--task` flag.
https://github.com/ultralytics/yolov5/blob/ac8691e20827ec6103c6f521397bb9f699ac8a52/test.py#L297

That being said, if your error is caused by 3rd party notebooks which we don't officially support then you should raise your issue directly with those authors. The official YOLOv5 notebook is here:
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb?hl=en

**Comment by fbagci** [comment]
Thank you for your response. 

Another thing I would like to ask is, is there a documentation of the methods you used while calculating the values in the confusion matrix? For example, suppose there are 50 images for a class in a dataset. Suppose we calculate the number of True positive samples for this class as 40. Here is why 10 examples were marked as wrong guesses. How does it compare predicted bounding box and ground truth bounding box? Thanks

**Comment by glenn-jocher** [comment]
@fbagci see ConfusionMatrix class in metrics.py:
https://github.com/ultralytics/yolov5/blob/ac8691e20827ec6103c6f521397bb9f699ac8a52/utils/metrics.py#L109-L182

**Comment by dadin852** [comment]
I wrote a very rough code which compares boxes. I'm not sure if  it fits your needs, but you can check it out!
[YOLO-Confusion-Matrix-with-custom-labels](https://github.com/dadin852/YOLO-Confusion-Matrix-with-custom-labels.git)

**Comment by glenn-jocher** [comment]
@dadin852 thanks for sharing your code! We recommend using the built-in evaluation tools in YOLOv5 which can be found in our documentation at https://docs.ultralytics.com/yolov5/. This provides standardized metrics and a clear interface for evaluating model performance.
